### ISSUE TITLE
`VkCreateBufferView` with depth-stencil formats assumes known texel size

### ISSUE THREAD

**Issue by Rua** [KhronosGroup/Vulkan-ValidationLayers#3225]
The `ValidateBufferViewRange` function checks for the size of a texel block using the `FormatElementSize` function. When given a depth-stencil format, this function returns a defined value, taken from the values in `kVkFormatTable`. But in the section on texel block sizes, the spec states:

> Depth/stencil formats are considered opaque and need not be stored in the exact number of bits per texel or component ordering indicated by the format enum.

This means that `FormatElementSize` can't return a known value for the size of these formats, since there isn't any. Implementations appear to be free to store more bytes than needed. This has further effects for checks that rely on the size, as is the case with `VkCreateBufferView`. One of the requirements (VUID-VkBufferViewCreateInfo-range-00929) is that the range must be a multiple of the format's texel block size. But in the case of depth-stencil formats, without a known size, this requirement appears to be impossible to meet.

**Comment by sfricke-samsung** [comment]
I agree the spec contradicts what `FormatElementSize` is doing, but for `VUID-VkBufferViewCreateInfo-range-00929` what device are you running on that supports `VK_FORMAT_FEATURE_UNIFORM_TEXEL_BUFFER_BIT` or `VK_FORMAT_FEATURE_STORAGE_TEXEL_BUFFER_BIT` for a depth/stencil format?

I was not aware that texel buffers in shaders could be depth/stencil format. I don't see why you would use something such as a `VK_FORMAT_D16_UNORM` over `VK_FORMAT_R16_UNORM` for a texel buffer in the first place

**Comment by Rua** [comment]
I'm not aware of any device that supports this. I suppose that this is a case of code that happens to do the right thing because of circumstance, but which could be potentially broken by a device implementer. As long as all users of `FormatElementSize` are aware that it gives wrong results when given a depth-stencil format, I guess there is no issue.

**Comment by sfricke-samsung** [comment]
so looking at all the uses of `FormatElementSize` I see it used for Multiplanar, Compressed, and Vertex attribute really only

So this explain probably why this hasn't surfaced a real issue yet.

Looking at this more, I think we could safely fix `FormatElementSize` to assert if depth/stencil format is used as it should never occur from what I am seeing as it would be misleading to return what is an opaque size

**Comment by ncesario-lunarg** [comment]
Just to clarify, checks such as https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/093e4ef78050220815d136548b80ad25961bb221/layers/buffer_validation.cpp#L1879 (which may be the only one, just happened to be the first assert I hit when I made this change) should then be skipped for depth/stencil formats?

**Comment by ncesario-lunarg** [comment]
Here's another one: https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/chap20.html#VUID-vkCmdCopyBufferToImage-pRegions-00171.

Is this VUID "invalid" for copy commands involving depth/stencil formats?

**Comment by sfricke-samsung** [comment]
ok, actually looking at this more I **think** the spec should go from saying

> Depth/stencil formats are considered opaque and need not be stored in the exact number of bits per texel

to something about how it still needs to support the **minimum needed**. For example, I would image a `D16` would **always** at least have 2 bytes per texel, but it **could** really be a D24 or D32 underneath

I plan to raise this up in the WG 

---

regardless, since `total_size` is only used for that warning, I see no harm in just wrapping that `maxResourceSize` warning around a `FormatIsColor` check

**Comment by sfricke-samsung** [comment]
> VUID-vkCmdCopyBufferToImage-pRegions-00171

Under `VkBufferImageCopy` in the spec it talks about depth/stencil copy 

like 

> data copied to or from the depth aspect of a VK_FORMAT_D16_UNORM or VK_FORMAT_D16_UNORM_S8_UINT format is tightly packed with one VK_FORMAT_D16_UNORM value per texel.

which I think I realized where maybe the original `FormatElementSize` depth values came from

**Comment by sfricke-samsung** [comment]
between the various PRs, small spec clarification, and now having the format util file generated from a proper XML scheme, this should have covered  all the issues here and this can be closed now IMO

**Comment by Rua** [comment]
What was the spec change that was made?

**Comment by sfricke-samsung** [comment]
2 Thing

1. For your original issue, there is no need to ever use depth/stencil as a size. So apps so use `R8` to mean 1 bytes, not `D8` for example. In the `Mandatory format support: depth/stencil` table a line at the end was add

> bufferFeatures must not support any features for these formats

and CTS coverage so make sure this case never occurs (the Validation Layers will throw an error if you try to use depth formats here)

2. There are plaes in the spec, such as normal copies, where a size is needed, but the Spec already has clear wording under the `VkBufferImageCopy` section and for those the Validation Layers were hardcoding the values, but now are using the information from the XML to correctly know the size of copies

```xml
<format name="VK_FORMAT_D16_UNORM_S8_UINT" class="D16S8" blockSize="3" texelsPerBlock="1">
            <component name="D" bits="16" numericFormat="UNORM"/>
            <component name="S" bits="8" numericFormat="UINT"/>
</format>
```

**Comment by spencer-lunarg** [comment]
closing for reasons as specified in the last comment
